packages/ipfs-core/src/components/pin/remote/client.js

```diff
@@ -0,0 +1,283 @@
+    return addrs.map(ma => {
+      const str = ma.toString()
+
+      // some relay-style transports add our peer id to the ma for us
+      // so don't double-add
+      if (str.endsWith(`/p2p/${peerId}`)) {
+        return str
+      }
+
+      return `${str}/p2p/${peerId}`
+    })
+  }
+
+  async function connectToDelegates (delegates) {
+    const addrs = delegates.map(multiaddr)
+    const promises = []
+    for (const addr of addrs) {
+      promises.push(swarm.connect(addr).catch(e => {
+        log('error connecting to pinning service delegate: ', e)
+      }))
+    }
+    await Promise.all(promises)
+  }
+
+  async function awaitPinCompletion (pinResponse) {
+    const pollIntervalMs = 100
+
+    let { status, requestid } = pinResponse
+    while (status !== 'pinned') {
+      if (status === 'failed') {
+        throw new Error('pin failed: ' + JSON.stringify(pinResponse.info))
+      }
+
+      await delay(pollIntervalMs)
+      const resp = await api.get(`/pins/${requestid}`)
+      pinResponse = await resp.json()
+      status = pinResponse.status
+    }
+
+    return formatPinResult(pinResponse.status, pinResponse.pin)
+  }
+
+  function formatPinResult (status, pin) {
+    const name = pin.name || ''
+    const cid = new CID(pin.cid)
+    return { status, name, cid }
+  }
+
+  /**
+   * Request that the remote service add a pin for the given CID.
+   *
+   * @param {CID|string} cid - CID to pin to remote service
+   * @param {AddOptions} options
+   *
+   * @returns {Promise<Pin>}
+   */
+  async function add (cid, options) {
+    const { background } = options
+    const name = options.name || ''
+    const origins = await originAddresses()
+    const addOpts = { cid: cid.toString(), name, origins }
+    const response = await api.post('/pins', { json: addOpts })
+    const responseBody = await response.json()
+
+    const { status, pin, delegates } = responseBody
+    connectToDelegates(delegates)
+
+    if (!background) {
+      return awaitPinCompletion(responseBody)
+    }
+
+    return formatPinResult(status, pin)
+  }
+
+  /**
+   * List pins from the remote service that match the given criteria. If no criteria are provided, returns all pins with the status 'pinned'.
+   *
+   * @param {Query} options
+   * @returns {AsyncGenerator<PinDetails>}
+   */
+  async function * _lsRaw (options) {
+    let status = options.status || []
+    if (status.length === 0) {
+      status = ['pinned']
+    }
+
+    const searchParams = new URLSearchParams()
+    if (options.name) {
+      searchParams.append('name', options.name)
+    }
+    for (const cid of (options.cid || [])) {
+      searchParams.append('cid', cid.toString())
+    }
+    for (const s of status) {
+      searchParams.append('status', s)
+    }
+
+    let resp = await api.get('/pins', { searchParams })
+    let body = await resp.json()
+    const total = body.count
+    let yielded = 0
+    while (true) {
+      if (body.results.length < 1) {
```

Comment:
This seems redundunt if it's empty next loop will be noon and next `if` clause will return and if doesn't it's probably better to fail.